### PR TITLE
[excel] (Shapes) An article for the upcoming Shape APIs

### PR DESCRIPTION
**Pull request by AlexJerabek**
This article covers shapes. It describes how to create, modify, and delete shapes of different types.

### CODE REVIEW

**Comment by o365devx** [comment]
OPS Build status updates of commit _[15ca759](https://github.com/OfficeDev/office-js-docs-pr/pull/837/commits/15ca759b052a44536754d74b1dda70cb15d16bfe)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/excel-add-ins-shapes.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-shapes.md) | :white_check_mark:Succeeded | |
[docs/excel/excel-add-ins-workbooks.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-workbooks.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C3%5C21%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201903212043094070-837%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=hMiqEKf01GETCM54agm%2Fr9KOOpnX4qGxgCr2TaSKh0o%3D&st=2019-03-21T20%3A39%3A31Z&se=2019-04-21T20%3A44%3A31Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by o365devx** [comment]
OPS Build status updates of commit _[beadf6f](https://github.com/OfficeDev/office-js-docs-pr/pull/837/commits/beadf6ffc5d34c5134b6765ea8d78f29864a061e)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/excel-add-ins-shapes.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-shapes.md) | :white_check_mark:Succeeded | |
[docs/excel/excel-add-ins-workbooks.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-workbooks.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C3%5C21%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201903212055064695-837%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=Fr5WHgnmK9a2eCZK0I1CYey4mZ0dEUv7BgdN0Ei%2BoOo%3D&st=2019-03-21T20%3A51%3A27Z&se=2019-04-21T20%3A56%3A27Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/excel-add-ins-shapes.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+Lines can be connected to other Shape objects. The `connectBeginShape` and `connectEndShape` methods attach the beginning and ending of a line to shapes at the specified connection points. The locations of these points vary by shape, but the `Shape.connectionSiteCount` can be used to ensure your add-in does not connect to a point that's out-of-bounds. A line is disconnected from any attached shapes using the `disconnectBeginShape` and `disconnectEndShape` methods.
+
+The following code sample connects the **"MyLine"** line to two shapes named **"LeftShape"** and **"RightShape"**.
+
+```js
+// This sample connects a line between two shapes at connection points '0' and '3'.
+Excel.run(function (context) {
+    var shapes = context.workbook.worksheets.getItem("MyWorksheet").shapes;
+    var line = shapes.getItem("MyLine").line;
+    line.connectBeginShape(shapes.getItem("LeftShape"), 0);
```
Comment (Rick-Kirkham) [inline docs/excel/excel-add-ins-shapes.md]: I can't run this in Script Lab. The methods Line.connectBeginShape and Line.connectEndShape do not seem to be in the beta d.ts.

File: docs/excel/excel-add-ins-shapes.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    line.connectEndShape(shapes.getItem("RightShape"), 3);
+    return context.sync();
+}).catch(errorHandlerFunction);
+```
+
+## Move and resize shapes
+
+Shapes sit on top of the worksheet. Their placement is defined by the `left` and `top` property. These act as margins from worksheet's respective edges, with [0, 0] being the upper-left corner. These can either be set directly or adjusted from their current position with the `incrementLeft` and `incrementTop` methods. How much a shape is rotated from the default position is also established in this manner, with the `rotation` property being the absolute amount and the `incrementRotation` method adjusting the existing rotation.
+
+A shape's depth relative to other shapes is defining by the `zorderPosition` property. This is set using the `setZOrder` method, which takes a [ShapeZOrder](/javascript/api/excel/excel.shapezorder). `setZOrder` adjusts the ordering of the current shape relative to the other shapes.
```
Comment (Rick-Kirkham) [inline docs/excel/excel-add-ins-shapes.md]: "defining" should be "defined"

File: docs/excel/excel-add-ins-shapes.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+Your add-in has a couple options for changing the height and width of shapes. Setting the `height` and `width` properties change that dimension without changing the other dimension. The `scaleHeight` and `scaleWidth` adjust the shape's respective dimensions relative to either the current or original size (based on the value of the provided [ShapeScaleType](/javascript/api/excel/excel.shapescaletype)). An optional [ShapeScaleFrom](/javascript/api/excel/excel.shapescalefrom) parameter specifies from where the shape scales (top-left corner, middle, or bottom-right corner). If the `lockAspectRatio` property is **true**, the scale methods maintain the shape's current aspect ratio by also adjusting the other dimension.
```
Comment (Rick-Kirkham) [inline docs/excel/excel-add-ins-shapes.md]:
```suggestion
Your add-in has a couple options for changing the height and width of shapes. Setting either the `height` or `width` property changes the specified dimension without changing the other dimension. The `scaleHeight` and `scaleWidth` adjust the shape's respective dimensions relative to either the current or original size (based on the value of the provided [ShapeScaleType](/javascript/api/excel/excel.shapescaletype)). An optional [ShapeScaleFrom](/javascript/api/excel/excel.shapescalefrom) parameter specifies from where the shape scales (top-left corner, middle, or bottom-right corner). If the `lockAspectRatio` property is **true**, the scale methods maintain the shape's current aspect ratio by also adjusting the other dimension.
```

File: docs/excel/excel-add-ins-shapes.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    textbox.height = 20;
+    textbox.width = 45;
+    textbox.name = "Textbox";
+    return context.sync();
+}).catch(errorHandlerFunction);
+```
+
+## Shape groups
+
+Shapes can be grouped together. This allows a user to treat them as a single entity. A [ShapeGroup](/javascript/api/excel/excel.shapegroup) is a type of `Shape`, so your add-in treats the group as a single shape.
```
Comment (Rick-Kirkham) [inline docs/excel/excel-add-ins-shapes.md]:
```suggestion
Shapes can be grouped together. This allows a user to treat them as a single entity for purposes of positioning, sizing, etc. A [ShapeGroup](/javascript/api/excel/excel.shapegroup) is a type of `Shape`, so your add-in treats the group as a single shape.
```

File: docs/excel/excel-add-ins-shapes.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+// This sample moves the previously created shape group to the right by 50 pixels.
+Excel.run(function (context) {
+    var shapes = context.workbook.worksheets.getItem("MyWorksheet").shapes;
+    var shapeGroup = sheet.shapes.getItem("Group");
+    shapeGroup.incrementLeft(50);
+    return context.sync();
+}).catch(errorHandlerFunction);
+```
+
+Individual shapes within the group are referenced through the `ShapeGroup.shapes` property, which is of type [GroupShapeCollection](/javascript/api/excel/excel.GroupShapeCollection). They are no longer accessible through the worksheet's shape collection after being grouped. As an example, if your worksheet had three shapes and they were all grouped together, the worksheet's `shapes.getCount` method would return a count of 1.
```
Comment (Rick-Kirkham) [inline docs/excel/excel-add-ins-shapes.md]: This paragraph seems like an important gotcha. I suggest putting it in an Important! note.

File: docs/excel/excel-add-ins-shapes.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+```js
+// This sample creates an image as a Shape object in the worksheet.
+var myFile = document.getElementById("selectedFile");
+var reader = new FileReader();
+
+reader.onload = (event) => {
+    Excel.run(function (context) {
+        var startIndex = event.target.result.indexOf("base64,");
+        var myBase64 = event.target.result.substr(startIndex + 7);
```
[the remark this answers is not in the document]
Comment (AlexJerabek) [reply docs/excel/excel-add-ins-shapes.md]: indexOf returns the index of the first character in "base64,", so the +7 is to get past those seven characters.

File: docs/excel/excel-add-ins-shapes.md (hunk 1 above)
Comment (AlexJerabek) [reply docs/excel/excel-add-ins-shapes.md]: Hmm, it's working for me (I have a sample in Script Lab that uses them). Those methods were renamed a couple weeks ago.  Please double check the libraries or use the SL sample and let me know if there's still problems.

**Comment by o365devx** [comment]
OPS Build status updates of commit _[a749e60](https://github.com/OfficeDev/office-js-docs-pr/pull/837/commits/a749e6096de9a2f12ac120cc2c4d9f2b384aaba4)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/excel-add-ins-shapes.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-shapes.md) | :white_check_mark:Succeeded | |
[docs/excel/excel-add-ins-workbooks.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-workbooks.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C3%5C21%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201903212335309094-837%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=ISk1SbVhwH%2BtDIdryLuGZx5iZBft3E4Zk7otu5S7Yc4%3D&st=2019-03-21T23%3A31%3A53Z&se=2019-04-21T23%3A36%3A53Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

**Comment by AlexJerabek** [comment]
Thanks @Rick-Kirkham. I responded to your feedback and incorporated all your suggestions in one way or another. Let me know if the SL sample isn't working for you.

**Comment by o365devx** [comment]
OPS Build status updates of commit _[f76e009](https://github.com/OfficeDev/office-js-docs-pr/pull/837/commits/f76e0092b72ef25924e9eb1fcbe01814aee0f316)_: 

### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[docs/excel/excel-add-ins-shapes.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-shapes.md) | :white_check_mark:Succeeded | |
[docs/excel/excel-add-ins-workbooks.md](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/excel/excel-add-ins-workbooks.md) | :white_check_mark:Succeeded | |
[docs/toc.yml](https://github.com/OfficeDev/office-js-docs-pr/blob/AlexJ-ShapeArticle/docs/toc.yml) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2019%5C3%5C21%5C57856ef7-e8cf-54ca-0d12-df5f42d1e8bb%5CPullRequest%5C201903212343067840-837%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=%2B1s%2BIk8UFyLD7WU1lkjIlQB8vB0DvBOYnviiUolqkNk%3D&st=2019-03-21T23%3A39%3A27Z&se=2019-04-21T23%3A44%3A27Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

File: docs/excel/excel-add-ins-shapes.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Rick-Kirkham) [reply docs/excel/excel-add-ins-shapes.md]: I get it. Thanks.

**Pull request merged**